JDK-8348030 : Extend Math.min and Math.max to support multiple parameters

aviad1486 opened the pull request:
This change extends the Math.min function to support multiple parameters, improving its usability and code readability.

Previously, finding the minimum value among multiple variables required using nested Math.min calls or converting the variables into an array and iterating through it. This enhancement provides a more intuitive and straightforward approach to achieve the same result.

Benefits:

Simplifies code by eliminating the need for nested Math.min calls.
Enhances readability, especially when comparing multiple values.
Offers consistency with existing Math.min usage patterns.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request matching fixVersion 25 to be approved (needs to be created)

### Issue
 * [JDK-8348030](https://bugs.openjdk.org/browse/JDK-8348030): Extend Math.min and Math.max to support multiple parameters (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23138/head:pull/23138` \
`$ git checkout pull/23138`

Update a local copy of the PR: \
`$ git checkout pull/23138` \
`$ git pull https://git.openjdk.org/jdk.git pull/23138/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23138`

View PR using the GUI difftool: \
`$ git pr show -t 23138`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23138.diff">https://git.openjdk.org/jdk/pull/23138.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23138#issuecomment-2599701048)
</details>

aviad1486 commented:
updating the commit message with relevant BUG-ID

minborg commented:
Was this proposal discussed before providing this PR?  Why is just `min()` proposed to change and not others like `max()` and `addExact()? This PR would require a [CSR](https://wiki.openjdk.org/display/csr/CSR+FAQs) to be filed.

aviad1486 commented:
Thank you for reviewing my PR and providing feedback.

I apologize for not discussing this enhancement on the mailing list prior to submitting the PR. My focus was on extending Math.min for multiple parameters to simplify code readability and reduce nested calls. 

Expanding Scope:
I chose to focus on min() for this initial contribution to keep the proposal small and manageable. However, I understand the value of symmetry in the API and would be open to extending similar functionality to **max()** and other related methods, such as addExact(), if the community agrees.

CSR Requirement:
Thank you for pointing out the need for a CSR. I would appreciate any guidance on how to file a CSR and the next steps to ensure that this enhancement aligns with OpenJDK’s standards and processes.

I look forward to your guidance and continuing to improve this contribution

aviad1486 commented:
I have updated the PR to include an extended Math.max function, similar to the changes made for Math.min. This ensures consistency in the API and provides a unified approach for handling multiple parameters across both methods. Additionally, I have updated the PR title to reflect these enhancements. Please let me know if further changes are required.

desruisseaux commented:
Aren't `Math.min(values)` methods duplicating `Arrays.stream(values).min()`? Except for the `float` type, for which there is no stream.

aviad1486 commented:
Thank you for your feedback!

While Arrays.stream(values).min() can indeed be used for int, long, and double, I believe that Math.min(values...) provides a more concise and intuitive alternative for these cases. Developers might prefer the simplicity of using a single method, especially when working with smaller sets of values where creating a stream might feel excessive.

As you pointed out, the Streams API does not support float arrays directly. The Math.min(float...) method addresses this gap, offering a consistent API for all primitive types.

Including varargs-based Math.min methods for all primitive types ensures consistency across the Math class and improves code readability.

I hope this clarifies the motivation behind the proposal. Please let me know if you have further suggestions or feedback!

minborg commented:
/csr

minborg commented:
Before dealing with a CSR, I suggest you socialize your idea/PR in the core-libs-dev@openjdk.org mailing list. You could reference this PR in the initial mail.

aviad1486 commented:
> Before dealing with a CSR, I suggest you socialize your idea/PR in the [core-libs-dev@openjdk.org](mailto:core-libs-dev@openjdk.org) mailing list. You could reference this PR in the initial mail.

Thanks for help, I'll do that

minborg commented:
Thanks for socializing your idea @aviad1486 ! 

I think we should *not* integrate this PR and there are many reasons for that including, but not limited to:

 * There are easy ways to provide this kind of functionality outside the JDK.
 * We are not sure there is a need for this in the community.
 * Adding these overloads means we commit to supporting them in perpetuity.
 * There are works ongoing to intensify `Math::min` and `Math::max`. This means we'd have to do the same here.
 * There are replacements for many similar constructs (e.g. using a reduction of streams).
 * The reviewing capacity is limited and currently, we are focusing on major additions to the Java platforms (e.g. Valhalla).
 * It is not clear which other methods in `Math` (and `MathExact`) should be expanded in a similar way.
 * It is not clear what performance, escape analysis, and inline factors this would impact.

Languages such as Haskel support "infix" operators meaning you could write something like `a 'min' b 'min' c`. If such a feature would be available in Java in the distant future, then the use of the proposed methods would be reduced but we would still have to maintain them forever.

So, I strongly suggest that we should close this PR without integrating it. Thanks again @aviad1486 for your effort.

aviad1486 commented:
Can you please keep the PR? I'm still working on it

בתאריך יום ו׳, 7 בפבר׳ 2025, 15:16, מאת Per-Ake Minborg ‏<
***@***.***>:

> Thanks for socializing your idea @aviad1486 <https://github.com/aviad1486>
> !
>
> I think we should *not* integrate this PR and there are many reasons for
> that including, but not limited to:
>
>    - There are easy ways to provide this kind of functionality outside
>    the JDK.
>    - We are not sure there is a need for this in the community.
>    - Adding these overloads means we commit to supporting them in
>    perpetuity.
>    - There are works ongoing to intensify Math::min and Math::max. This
>    means we'd have to do the same here.
>    - There are replacements for many similar constructs (e.g. using a
>    reduction of streams).
>    - The reviewing capacity is limited and currently, we are focusing on
>    major additions to the Java platforms (e.g. Valhalla).
>    - It is not clear which other methods in Math (and MathExact) should
>    be expanded in a similar way.
>    - It is not clear what performance, escape analysis, and inline
>    factors this would impact.
>
> Languages such as Haskel support "infix" operators meaning you could write
> something like a 'min' b 'min' c. If such a feature would be available in
> Java in the distant future, then the use of the proposed methods would be
> reduced but we would still have to maintain them forever.
>
> So, I strongly suggest that we should close this PR without integrating
> it. Thanks again @aviad1486 <https://github.com/aviad1486> for your
> effort.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/23138#issuecomment-2642878666>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEHBFNK742B4HJSYSAB4YQL2OSW2ZAVCNFSM6AAAAABVHL4VJOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBSHA3TQNRWGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
